### PR TITLE
Harrison/matching engine vectorstore

### PR DESCRIPTION
**Pull request by hwchase17**

### CODE REVIEW

**Comment by hwchase17** [comment]
@tomaspiaggio i tried to update this and get it to pass lint - how does it look?

File: langchain/vectorstores/matching_engine.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+        planning to do a real time implementation. While reading is a real time
+        operation, updating the index takes close to one hour.
+
+        Attributes:
+            project_id: The GCS project id.
+            index: The created index class. See
+            ~:func:`MatchingEngine.from_components`.
+            endpoint: The created endpoint class. See
+            ~:func:`MatchingEngine.from_components`.
+            gcs_client: The Google Cloud Storage client.
```
Comment (tomaspiaggio) [inline langchain/vectorstores/matching_engine.py:69]: @hwchase17 maybe the attributes order should match the constructor arguments? Also I think `gcs_bucket_name` and `credentials` are missing.

File: langchain/vectorstores/matching_engine.py [hunk 2]
```diff
@@ -0,0 +1,433 @@
...
+    @classmethod
+    def from_texts(
+        cls: Type["MatchingEngine"],
+        texts: List[str],
+        embedding: Embeddings,
+        metadatas: Optional[List[dict]] = None,
+        **kwargs: Any,
+    ) -> "MatchingEngine":
+        """Use from components instead."""
+        raise NotImplementedError(
```
Comment (tomaspiaggio) [inline langchain/vectorstores/matching_engine.py:237]: @hwchase17 why was this removed? Previously, it would have worked regardless if the developer called `from_texts` or `from_components`

File: langchain/vectorstores/matching_engine.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+        return credentials
+
+    @classmethod
+    def _create_index_by_id(
+        cls, index_id: str, project_id: str, region: str, credentials: "Credentials"
+    ) -> MatchingEngineIndex:
+        """Creates a MatchingEngineIndex object by id.
+
+        Args:
+            index_id: The created index id.
```
Comment (tomaspiaggio) [inline langchain/vectorstores/matching_engine.py:343]: @hwchase17 Add other arguments to `Args`

File: langchain/vectorstores/matching_engine.py [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+        )
+
+    @classmethod
+    def _create_endpoint_by_id(
+        cls, endpoint_id: str, project_id: str, region: str, credentials: "Credentials"
+    ) -> MatchingEngineIndexEndpoint:
+        """Creates a MatchingEngineIndexEndpoint object by id.
+
+        Args:
+            endpoint_id: The created endpoint id.
```
Comment (tomaspiaggio) [inline langchain/vectorstores/matching_engine.py:366]: @hwchase17 same as above, add other arguments to `Args`

File: langchain/vectorstores/matching_engine.py [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+        Returns:
+            A configured MatchingEngineIndexEndpoint.
+            :param project_id:
```
Comment (tomaspiaggio) [inline langchain/vectorstores/matching_engine.py:370]: @hwchase17 I think this moved

File: langchain/vectorstores/matching_engine.py [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+        aiplatform.init(
+            project=project_id,
+            location=region,
+            staging_bucket=gcs_bucket_name,
+            credentials=credentials,
+        )
+
+    @classmethod
+    def _get_default_embeddings(cls) -> TensorflowHubEmbeddings:
+        """This function returns the default embedding."""
```
Comment (tomaspiaggio) [inline langchain/vectorstores/matching_engine.py:432]: @hwchase17 Add `Returns:` to comment as other methods and functions.

**Review by tomaspiaggio — CHANGES_REQUESTED** [review]
@hwchase17 some minor details. Otherwise, it lgtm. lmk if you need anything else from me!

**Comment by Jflick58** [comment]
@hwchase17 really interested in seeing this get merged! Open to me fixing tomaspiaggio's comments?

File: langchain/vectorstores/matching_engine.py (hunk 2 above)
Comment (dev2049) [reply langchain/vectorstores/matching_engine.py:237]: think the from_texts interface doesn't fit super well, since it only allows texts and embedding as args, whereas the init and from_components both expect a number of args. we could technically pipe all kwargs for from_text into one of those two but it's pretty hacky. also there's no real value to using from_texts for this class, since it'd just be wrapping instantiation and calling add_texts

File: docs/modules/indexes/vectorstores/examples/matchingengine.ipynb [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+    "texts = ['The cat sat on', 'the mat.', 'I like to', 'eat pizza for', 'dinner.', 'The sun sets', 'in the west.']\n",
+    "\n",
+    "\n",
+    "vector_store = MatchingEngine.from_components(\n",
+    "    texts=texts,\n",
+    "    project_id=\"<my_project_id>\",\n",
+    "    region=\"<my_region>\",\n",
+    "    gcs_bucket_uri=\"<my_gcs_bucket>\",\n",
+    "    index_id=\"<my_matching_engine_index_id>\",\n",
+    "    enpoint_id=\"<my_matching_engine_endpoint_id>\"\n",
```
Comment (RajeshThallam) [inline docs/modules/indexes/vectorstores/examples/matchingengine.ipynb:51]: Typo - It should `endpoint_id`

**Review by tomaspiaggio — APPROVED** [review]
